File: web/default/src/features/security/api/security.ts
import { apiClient } from '@/lib/api-client'

export interface SecurityGroup {
  id: number
  name: string
  description: string
  status: number
  parent_id: number
  depth: number
  path: string
  sort_order: number
  created_at: number
  updated_at: number
}

export interface SecurityRule {
  id: number
  group_id: number
  group_name?: string
  name: string
  type: number
  content: string
  extra_config: string
  action: number
  priority: number
  risk_score: number
  status: number
  created_at: number
  updated_at: number
}

export interface SecurityPolicy {
  id: number
  user_id: number
  user_name?: string
  group_id: number
  group_name?: string
  scope: number
  default_action: number
  custom_response: string
  whitelist_ips: string
  status: number
  created_at: number
  updated_at: number
}

export interface SecurityHitLog {
  id: number
  request_id: string
  user_id: number
  user_name?: string
  model_name: string
  action: number
  risk_level: number
  risk_score: number
  original_content_hash: string
  processed_content?: string
  match_detail?: string
  ip: string
  created_at: number
}

export interface DashboardData {
  summary: {
    total_detections: number
    total_interceptions: number
    total_alerts: number
    today_detections: number
  }
  top_categories: Array<{ category: string; count: number }>
  top_users: Array<{ user_id: number; user_name: string; count: number }>
  top_models: Array<{ model_name: string; count: number }>
  risk_distribution: { low: number; medium: number; high: number; critical: number }
}

export const securityApi = {
  // Groups
  getGroups: (params?: { page?: number; page_size?: number; status?: number; parent_id?: number }) =>
    apiClient.get('/api/security/groups', { params }).then((r) => r.data),
  createGroup: (data: Partial<SecurityGroup>) => apiClient.post('/api/security/groups', data).then((r) => r.data),
  updateGroup: (id: number, data: Partial<SecurityGroup>) =>
    apiClient.put(`/api/security/groups/${id}`, data).then((r) => r.data),
  deleteGroup: (id: number) => apiClient.delete(`/api/security/groups/${id}`).then((r) => r.data),
  copyGroup: (id: number) => apiClient.post(`/api/security/groups/${id}/copy`).then((r) => r.data),

  // Rules
  getRules: (params?: { page?: number; page_size?: number; group_id?: number; type?: number; status?: number }) =>
    apiClient.get('/api/security/rules', { params }).then((r) => r.data),
  createRule: (data: Partial<SecurityRule>) => apiClient.post('/api/security/rules', data).then((r) => r.data),
  updateRule: (id: number, data: Partial<SecurityRule>) =>
    apiClient.put(`/api/security/rules/${id}`, data).then((r) => r.data),
  deleteRule: (id: number) => apiClient.delete(`/api/security/rules/${id}`).then((r) => r.data),

  // Policies
  getPolicies: (params?: { page?: number; page_size?: number; user_id?: number; status?: number }) =>
    apiClient.get('/api/security/policies', { params }).then((r) => r.data),
  createPolicy: (data: Partial<SecurityPolicy>) => apiClient.post('/api/security/policies', data).then((r) => r.data),
  updatePolicy: (id: number, data: Partial<SecurityPolicy>) =>
    apiClient.put(`/api/security/policies/${id}`, data).then((r) => r.data),
  deletePolicy: (id: number) => apiClient.delete(`/api/security/policies/${id}`).then((r) => r.data),

  // Logs
  getLogs: (params?: { page?: number; page_size?: number; user_id?: number; action?: number; risk_level?: number }) =>
    apiClient.get('/api/security/logs', { params }).then((r) => r.data),

  // Dashboard
  getDashboard: (params?: { start_time?: number; end_time?: number }) =>
    apiClient.get('/api/security/dashboard', { params }).then((r) => r.data),
}

File: dto/security.go
package dto

import (
	"time"
)

// ========== Security Group DTOs ==========

type SecurityGroupRequest struct {
	Name        string `json:"name" binding:"required,max=128"`
	Description string `json:"description" binding:"max=255"`
	ParentID    int64  `json:"parent_id"`
	SortOrder   int    `json:"sort_order"`
}

type SecurityGroupResponse struct {
	ID          int64     `json:"id"`
	Name        string    `json:"name"`
	Description string    `json:"description"`
	Status      int       `json:"status"`
	ParentID    int64     `json:"parent_id"`
	Depth       int       `json:"depth"`
	Path        string    `json:"path"`
	SortOrder   int       `json:"sort_order"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// ========== Security Rule DTOs ==========

type SecurityRuleRequest struct {
	GroupID     int64  `json:"group_id" binding:"required"`
	Name        string `json:"name" binding:"required,max=128"`
	Type        int    `json:"type" binding:"required,oneof=1 2 3 4"`
	Content     string `json:"content" binding:"required"`
	ExtraConfig string `json:"extra_config"`
	Action      int    `json:"action" binding:"required,oneof=1 2 3 4 5"`
	Priority    int    `json:"priority"`
	RiskScore   int    `json:"risk_score" binding:"min=0,max=100"`
}

type SecurityRuleResponse struct {
	ID          int64     `json:"id"`
	GroupID     int64     `json:"group_id"`
	GroupName   string    `json:"group_name,omitempty"`
	Name        string    `json:"name"`
	Type        int       `json:"type"`
	Content     string    `json:"content"`
	ExtraConfig string    `json:"extra_config"`
	Action      int       `json:"action"`
	Priority    int       `json:"priority"`
	RiskScore   int       `json:"risk_score"`
	Status      int       `json:"status"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// ========== Security Policy DTOs ==========

type SecurityPolicyRequest struct {
	UserID         int64  `json:"user_id" binding:"required"`
	GroupID        int64  `json:"group_id" binding:"required"`
	Scope          int    `json:"scope" binding:"oneof=1 2 3"`
	DefaultAction  int    `json:"default_action" binding:"oneof=1 2 3 4 5"`
	CustomResponse string `json:"custom_response"`
	WhitelistIPs   string `json:"whitelist_ips"`
}

type SecurityPolicyResponse struct {
	ID             int64     `json:"id"`
	UserID         int       `json:"user_id"`
	UserName       string    `json:"user_name,omitempty"`
	GroupID        int64     `json:"group_id"`
	GroupName      string    `json:"group_name,omitempty"`
	Scope          int       `json:"scope"`
	DefaultAction  int       `json:"default_action"`
	CustomResponse string    `json:"custom_response"`
	WhitelistIPs   string    `json:"whitelist_ips"`
	Status         int       `json:"status"`
	CreatedAt      time.Time `json:"created_at"`
	UpdatedAt      time.Time `json:"updated_at"`
}

// ========== Security Hit Log DTOs ==========

type SecurityHitLogQuery struct {
	Page       int    `form:"page"`
	PageSize   int    `form:"page_size"`
	UserID     int    `form:"user_id"`
	ModelName  string `form:"model_name"`
	StartTime  int64  `form:"start_time"`
	EndTime    int64  `form:"end_time"`
	GroupID    int64  `form:"group_id"`
	Action     int    `form:"action"`
	RiskLevel  int    `form:"risk_level"`
	ContentType int   `form:"content_type"`
}

type SecurityHitLogResponse struct {
	ID                  int64     `json:"id"`
	RequestID           string    `json:"request_id"`
	UserID              int       `json:"user_id"`
	UserName            string    `json:"user_name,omitempty"`
	ChannelID           int       `json:"channel_id"`
	ModelName           string    `json:"model_name"`
	TokenID             int       `json:"token_id"`
	RuleID              int64     `json:"rule_id"`
	RuleName            string    `json:"rule_name,omitempty"`
	GroupID             int64     `json:"group_id"`
	GroupName           string    `json:"group_name,omitempty"`
	ContentType         int       `json:"content_type"`
	Action              int       `json:"action"`
	RiskLevel           int       `json:"risk_level"`
	RiskScore           int       `json:"risk_score"`
	OriginalContentHash string    `json:"original_content_hash"`
	ProcessedContent    string    `json:"processed_content,omitempty"`
	MatchDetail         string    `json:"match_detail,omitempty"`
	IP                  string    `json:"ip"`
	CreatedAt           time.Time `json:"created_at"`
}

// ========== Security Dashboard DTOs ==========

type SecurityDashboardRequest struct {
	StartTime int64 `form:"start_time"`
	EndTime   int64 `form:"end_time"`
}

type SecurityDashboardResponse struct {
	Summary struct {
		TotalDetections    int `json:"total_detections"`
		TotalInterceptions int `json:"total_interceptions"`
		TotalAlerts        int `json:"total_alerts"`
		TodayDetections    int `json:"today_detections"`
	} `json:"summary"`
	TopCategories []struct {
		Category string `json:"category"`
		Count    int    `json:"count"`
	} `json:"top_categories"`
	TopUsers []struct {
		UserID   int    `json:"user_id"`
		UserName string `json:"user_name"`
		Count    int    `json:"count"`
	} `json:"top_users"`
	TopModels []struct {
		ModelName string `json:"model_name"`
		Count     int    `json:"count"`
	} `json:"top_models"`
	RiskDistribution struct {
		Low      int `json:"low"`
		Medium   int `json:"medium"`
		High     int `json:"high"`
		Critical int `json:"critical"`
	} `json:"risk_distribution"`
}

// ========== Security Check DTOs ==========

type SecurityCheckRequest struct {
	UserID    int    `json:"user_id" binding:"required"`
	Content   string `json:"content" binding:"required"`
	ModelName string `json:"model_name"`
}

type SecurityCheckResponse struct {
	Detected        bool                   `json:"detected"`
	Action          int                    `json:"action"`
	ActionName      string                 `json:"action_name"`
	RiskScore       int                    `json:"risk_score"`
	RiskLevel       int                    `json:"risk_level"`
	ProcessedContent string                `json:"processed_content,omitempty"`
	Matches         []SecurityMatchResult  `json:"matches,omitempty"`
}

type SecurityMatchResult struct {
	RuleID      int64  `json:"rule_id"`
	GroupID     int64  `json:"group_id"`
	Type        int    `json:"type"`
	MatchedText string `json:"matched_text"`
	Position    [2]int `json:"position"`
}

// ========== Common Response ==========

type SecurityStatusResponse struct {
	Enabled      bool  `json:"enabled"`
	RuleCount    int64 `json:"rule_count"`
	GroupCount   int64 `json:"group_count"`
	PolicyCount  int64 `json:"policy_count"`
	CacheEnabled bool  `json:"cache_enabled"`
}
